readme.md
-------

# ENHANCED PM SYSTEM v3.5.2

[**By Dougiefresh**](http://www.simplemachines.org/community/index.php?action=profile;u=253913) - [Link to Mod](http://custom.simplemachines.org/mods/index.php?mod=3736)

-------

# Introduction
This mod makes some changes to the PM system, seperating the PM conversations from the PM list.  It forces all users to conversation mode and makes users click on the message they want to view.  This keeps the forum from marking the latest message as read if you glance over the message list and click on something else and want to read a message that isn't the latest message received.

It adds an Outbox to the mix and seperates unread PMs from those that have been read.  It changes the "labels" into "folders" and removes the contents of the PM from the email notification.  It will allow users to edit PMs that they have sent out, as long as NO RECIPIENT has read it.  It also allows users to unsend and delete any unread PMs to other users.

This mod adds settings in the "Features and Options" page.  By default, this mod forces the view for personal messages to Conversation mode, but there is a new switch permitting the users to select a different mode other than Conversation mode.  The "PM Remove Inbox Label" and "Copy to Outbox" options are removed from the Personal Messages Settings page and set as default.

Version 2.0 sports an enhanced PM search system, which allows you to search Sent Items and the Outbox for PMs you specify.  It also allows the user to search them by username, just like the core functionality provided for the Inbox.  You are allowed to move PMs to folders for Inbox PMs, and allowed to delete PMs from all folders.

Version 3.0 adds the PM Quick Reply box, as well as changes some options to membergroup permissions....

As a anti-spam measure, this mod can block new members from sending PMs during the 24-hour period after their account is activated.  Please note that members activated before this mod is installed will not be affected by this anti-spam measure, as the activation time is not recorded in the database prior to this mod.

## Related Discussions

- [DONE] [PM Separate List and View Request](http://www.simplemachines.org/community/index.php?topic=420396)
- [DONE] [Always PM Conversation and Save a Copy](http://www.simplemachines.org/community/index.php?topic=427762)
- [DONE] [Private Message Email Notification](http://www.simplemachines.org/community/index.php?topic=426840)
- [DONE] [PM outbox/sent like phpbb](http://www.simplemachines.org/community/index.php?topic=128571)
- [DONE] [Unsend PM](http://www.simplemachines.org/community/index.php?topic=29990)
- [DONE] [Outbox with editable PMs](http://www.simplemachines.org/community/index.php?topic=403840)
- [DONE] [Require Subject for PMs](http://www.simplemachines.org/community/index.php?topic=333117)
- [DONE] [Block PM on the first day](http://www.simplemachines.org/community/index.php?topic=460098)
- [DONE] **NEW FOR v2.0:** [Search by Sent PMs](http://www.simplemachines.org/community/index.php?topic=498600)
- [DONE] **NEW FOR v2.0:** [Perform Actions on Searched-For PMs](http://www.simplemachines.org/community/index.php?topic=521508.msg3713864#msg3713864)
- [DONE] **NEW FOR V2.0:** ["User xx has blocked your personal message."](http://www.simplemachines.org/community/index.php?topic=478886)
- [DONE] **NEW FOR V2.1:** [Allow Members to Send PM's While Muted](http://www.simplemachines.org/community/index.php?topic=412263)
- [DONE] **NEW FOR V2.1:** [PM Message Rules - Members can avoid Admin PMs & Warning PMs](http://www.simplemachines.org/community/index.php?topic=412757)
- [DONE] **NEW FOR V2.2:** [A "Quote to All" in PM Conversation mode?](http://www.simplemachines.org/community/index.php?topic=503208.msg3539234#msg3539234)
- [DONE] **NEW FOR V3.0:** [PM Quick Reply](http://www.simplemachines.org/community/index.php?topic=521508.msg3804931#msg3804931)
- [INCOMPLETE] [Move Personal Messages](http://www.simplemachines.org/community/index.php?topic=295253)

## Permission Notices

- [MLM](http://www.simplemachines.org/community/index.php?action=profile;u=261314) gave permission to use his changes from [Always PM Conversation and Save a copy](http://www.simplemachines.org/community/index.php?topic=427762) discussion.
- [floydpink](http://www.simplemachines.org/community/index.php?action=profile;u=73143) gave permission to use his changes from [Move Personal Messages](http://www.simplemachines.org/community/index.php?topic=295253) discussion.

## Mail Icons

- [Ultimate Gnome](http://code.google.com/p/ultimate-gnome/) iconset (released under a GPL license) found at [IconFinder.com](http://www.iconfinder.com/search/?q=mail+iconset%3AUltimateGnome)

## Compatibility Notes
This mod was tested on SMF 2.0.9, but should work on SMF 2.1 Beta 2, as well as SMF 2.0 and up.  SMF 1.x is not and will not be supported.

The [PM Attachments](http://custom.simplemachines.org/mods/index.php?mod=1974) mod should be installed first to avoid any conflict with this mod.

## Questions and Answers
**Q:** Will this mod ever include PM attachment support?

**A:** No, there is a [PM Attachments](http://custom.simplemachines.org/mods/index.php?mod=1974) mod already developed that supports this.  The exception is editing PMs with attachments, which this mod will eventually support fully.

**Q:** Why can't I edit the attachments on a PM?

**A:** At present time, it is not possible to list and/or remove files attached to the PM while editing the PM.  Future versions of this mod will resolve this issue.

**Q:** Why can't I move messages from the Outbox and/or Sent Items folder?

**A:** At present time, it is not possible to mass-move messages from the Outbox or Sent Items folder.  Future versions of this mod may resolve this issue.

## Translators

- [@rjen](http://www.simplemachines.org/community/index.php?action=profile;u=287786) - Dutch translation
- [messire_Leonard](https://www.simplemachines.org/community/index.php?action=profile;u=431601) - Russian translation
- [Rock Lee](https://www.simplemachines.org/community/index.php?action=profile;u=322597) - Spanish Latin translation

## Changelog
The changelog has been removed and can be seen at [XPtsp.com](http://www.xptsp.com/board/index.php?topic=8.msg133#msg133).

## License
Copyright (c) 2015 - 2018, Douglas Orend
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Changelog.txt
v3.5.2 - October 12th, 2018
o Support for SMF 2.0.15 (by aegersz)

v3.5.1 - August 19th, 2016
o Added Russian translation to the mod.  Thanks, [url=https://www.simplemachines.org/community/index.php?action=profile;u=431601]messire_Leonard[/url]!
o Added Spanish Latin translation to the mod.  Thanks, [url=https://www.simplemachines.org/community/index.php?action=profile;u=322597]Rock Lee[/url]!

v3.5 - December 28th, 2015
o Added Spanish translation to the mod.  Thanks, [url=http://www.simplemachines.org/community/index.php?action=profile;u=422971]d3vcho[/url]!
o Fixed installation issue with UTF8 Brazilian Portuguese language file.

v3.4 - December 2nd, 2015
o Fixed issue where searching for sent items resulted in error message.
o Fixed several undefined variable error messages in PersonalMessage.php.
o Fixed permission to allow member to send PMs 24 hours after registering.

v3.3 - September 2nd, 2015
o Removed support for SMF 2.1 Beta 1
o Added support for SMF 2.1 Beta 2

v3.2 - August 7th, 2015
o Modifications made so that it can be installed beside "Slash through Banned Usernames" mod.
o Changed "most recent personal message" into "most recent conversation" on PM settings page.
o Added "Show most recent personal messages in conversation at top" to PM settings page.

v3.1.1 - July 6th, 2015
o Added Dutch translation, courtsey of [url=http://www.simplemachines.org/community/index.php?action=profile;u=287786]@rjen[/url]

v3.1 - June 8th, 2015
o Added additional restriction so PM Quick Reply isn't shown if no PMs are present.
o Fixed undefined index error in [b]PersonalMessage.template.php[/b] for SMF 2.0.

v3.0 - June 5th, 2015
o Updated for SMF 2.1 Beta 1.
o Added dropdown box for PM quick reply options in the PM Settings page.
o Added PM quick reply form to the PM discussion page.
o Added missing checkbox for option to disable blocking PMs from admin.
o Seperated the settings into their own page under Modifications => Enhanced PM.
o Converted "Deny Edit", "Deny Unsend" and "Block 1st Day" checkboxes into group-based permissions.
o Moved CSS modifications into their own CSS file.
o Moved language modifications into their own language file.
o Incorporated bug fix and quote-to-all functionality change suggested by "Geor'G". ( http://www.simplemachines.org/community/index.php?topic=521508.msg3812946#msg3812946 )
o Fixed some undefined variable/element errors in the "pretty-fied" Search PM area.
o Reinserted the "Move PMs into folders" operation back into the SMF 2.0 install file.
o NOTE: Upgrade not possible from version 2.8 because of language and CSS file changes.

v2.8 - March 5th, 2015
o Fixed bug where element is used before checking to see if it has been defined yet.
o Unfixed "bug" in the Manage Rules page that allowed users to assign more than one action to a Rule.
o Unfixed "bug" in function ApplyRules that allowed one PM to be put in multiple folders.
o POLICY CHANGE: Upgrade only possible from version 2.6 and 2.7.

v2.7 - January 16th, 2015
o License change to Simplified BSD.  No functionality change.

v2.6 - November 9th, 2014
o Fixed issue where this mod's language strings were put in the wrong file.

v2.5 - October 19th, 2014
o Another bug fix to resolve PM display mode not being set to conversation correctly

v2.4 - October 18th, 2014
o Mod rearranged due to some servers having difficulty processing this mod correctly...
o Upgrade files now correctly reflect the modifications between each version....
o 2nd attempt to fix bug that prevents latest message from being shown if more has been added to convo...

v2.3 - September 14th, 2014
o Mod upgradable from v2.x to v2.3.
o Fixed bugs that prevented latest message from being shown if more has been added to the conversation.
o Fixed install bug so that v2.2 operation in core/PersonalMessage.template.php is now ignored if doesn't exist.
o Fixed uninstaller from v2.2 because it doesn't include reversing the upgrade_2.2.xml instructions.

v2.2 - September 9th, 2014
o Mod now upgradable from version 2.0 and 2.1 to version 2.2!
o "Quote to All" feature implemented, requested by SilverWings.
o Fixed a bug where no Search button was shown while searching when no labels were defined, reported by Jeff B.
o More bug fixes where the PM display mode not being set to conversation correctly, reported by SilverWings.
o Fixed a bug that marks PMs as read even though user hasn't opened the PM yet, reported by ExSpirit
o Re-Added code that v2.1 removed by mistake which removes default subject from PMs (noticed during rebuild)
o Fixed a template bug in the Core theme's PersonalMessage.template.php.

v2.1 - August 24th, 2014
o Fixed a bug where a PM message is read by the user and still marked NEW, reported by ExSpirit
o Fixed a bug where user info is not updated when sent a new PM, reported by ExSpirit
o Fixed a bug where the PM display mode not being set to conversation correctly, reported by SilverWings
o Fixed a bug where PMs deleted from Outbox are incorrectly deleted, reported by SilverWings
o Added option to allow muted members the ability to send PMs.  Thanks, Corelogik!
o Added option to allow members to not block Admin PMs within the rules.  Thanks, danielwmoore!

v2.0 - July 9th, 2014
o If user is on recipient's PM ignore list, it is shown upon entering the reply page.
o Fix for duplicate buttons showing up for PMs in Core theme.
o Added ability to mass-delete of PMs in the Outbox.
o Added support for searching Sent Items and Outbox folders for PMs
o Search results now indicates whether the PMs found have been read or not.
o Copied PM display code for use by complete search result template...
o Added ability to mass-delete from all folders during searches.
o Added ability to mass-move messages from the Inbox and Folders during searches.
o Added ability to search Sent Items and Outbox PMs by username.
o Minor reorganization of template code to avoid duplication of necessary functions.

v1.6 - July 2nd, 2014

EnhancedPMSystem.english.php
<?php
/**********************************************************************************
* EnhancedPMSystem.english.php - English language file of the Split Forum Mod
*********************************************************************************
* This program is distributed in the hope that it is and will be useful, but
* WITHOUT ANY WARRANTIES; without even any implied warranty of MERCHANTABILITY
* or FITNESS FOR A PARTICULAR PURPOSE.
*********************************************************************************
* This work is licensed under a Creative Commons Attribution 3.0 Unported License
**********************************************************************************/
if (!defined('SMF'))
	die('Hacking attempt...');

// Replacement language strings for the Enhanced PM System:
//==================================================================
$txt['quote_to_all'] = 'Quote to All';
$txt['pm_unread'] = 'Unread';
$txt['pm_msg_label_title'] = 'Move Message';
$txt['pm_msg_label_apply'] = 'Move to';
$txt['pm_msg_label_remove'] = 'Current Folder';
$txt['pm_sel_label_title'] = 'Move Selected';
$txt['pm_manage_labels'] = 'Manage Folders';
$txt['pm_labels_delete'] = 'Are you sure you wish to delete the selected folders?';
$txt['pm_labels_desc'] = 'From here you can add, edit and delete the folders used in your personal message center.';
$txt['pm_label_add_new'] = 'Add New Folder';
$txt['pm_label_name'] = 'Folder Name';
$txt['pm_labels_no_exist'] = 'You currently have no folders setup!';
$txt['pm_labels'] = 'Folders';
$txt['pm_search_choose_label'] = 'Choose folders to search by, or search all';
$txt['pm_email'] = 'You have just been sent a personal message by SENDER on ' . $context['forum_name'] . '.' . "\n\n" . 'IMPORTANT: Remember, this is just a notification. Please do not reply to this email.';
$txt['unread_items'] = 'Outbox';
$txt['unsend_item'] = 'Unsend';
$txt['pm_rule_label'] = 'Move message to folder';
$txt['pm_rule_sel_label'] = 'Select Folder';

$txt['pm_edit'] = 'Edit';
$txt['edit_message'] = 'Edit Message';
$txt['pm_unsent_denied'] = 'You cannot unsend a message.';
$txt['pm_unsent'] = 'Your message has been unsent from all recipients successfully.';
$txt['pm_unsent_some'] = 'Your message has been unsent from all recipients who haven\'t read it successfully.';
$txt['pm_edited'] = 'Your message has been edited and all new recipients have been sent emails.';
$txt['pm_no_edit'] = 'Unable to edit PM because at least one recipient has read the PM that you are wanting to edit.';
$txt['pm_send_blocked'] = 'You are not allowed to send PMs until 24 hours after account activation.';
$txt['unsend_message'] = 'Unsend this message to all recipients who haven\'t read it yet';
$txt['pm_search_what'] = 'Search group';
$txt['pm_search_to_user'] = 'To user';

$txt['pm_actions_filter_by_label'] = 'Filter By Folder';
$txt['pm_current_label'] = 'Folder';
$txt['pm_readable_label'] = 'move to folder &quot;{LABEL}&quot;';

// Language string for PM Quick Reply:
//==================================================================
$txt['quick_reply_desc'] = 'With Quick-Reply you can write a PM when viewing a conversation without loading a new page. You can still use bulletin board code and smileys as you would in a normal PM.';

// Language strings for the Enhanced PM System settings page:
//==================================================================
$txt['eps_area'] = 'Enhanced PMs';
$txt['eps_title'] = 'Enhanced PM System';
$txt['eps_new_version'] = 'Enhanced PM System mod version %s is available for download!';
$txt['eps_no_update'] = 'Your install of Enhanced PM System mod is up to date!';
$txt['eps_allow_unsend'] = 'Membergroups allowed to Unsend an unread PM';
$txt['eps_allow_edit'] = 'Membergroups allowed to Edit an unread PM';
$txt['eps_block_1st_day'] = 'Membergroups that can\'t PM for the first day';
$txt['eps_pm_view_switch'] = 'Allow users to select PM views other than conversation mode?';
$txt['eps_send_pm_while_muted'] = 'Allow muted members to send PMs?';
$txt['eps_allow_block_admin'] = 'Allow members to block Admin PMs in PM rules?';

// Language strings for the Enhanced PM System permissions:
//==================================================================
$txt['permissionname_eps_block_1st_day'] = '<strong>NOT</strong> allowed to PM on 1st day';
$txt['permissionhelp_eps_block_1st_day'] = 'This setting stops members of this group from sending any PMs within the 1st day (24 hours) of their membership.  Requires the \'Read personal messages\' permission.';
$txt['permissionhelp_eps_allow_edit'] = 'This setting allows members of this group to edit personal messages sent to others, as long as the message hasn\'t been read by any recipient.  Requires the \'Read personal messages\' permission.';
$txt['permissionname_eps_allow_edit'] = 'Allowed to Edit unread PMs';
$txt['permissionhelp_eps_allow_unsend'] = 'This setting allows members of this group to unsend (or delete) personal messages sent to others, as long as the message hasn\'t been read by any recipient.  Does not affect emails already sent regarding PM.  Requires the \'Read personal messages\' permission.';
$txt['permissionname_eps_allow_unsend'] = 'Allowed to Unsend unread PMs';

// Language strings for the Enhanced PM System profile area:
//==================================================================
$txt['display_pm_quick_reply'] = 'Use quick reply on PM display:';
if (!isset($txt['display_quick_reply1']))
{
	$txt['display_quick_reply1'] = 'don\'t show at all';
	$txt['display_quick_reply2'] = 'show, off by default';
	$txt['display_quick_reply3'] = 'show, on by default';
}
$txt['recent_pms_at_top'] = 'Show most recent conversation at top of list.';
$txt['recent_msgs_at_top'] = 'Show most recent personal messages in conversation at top.';

?>